packages/salesforcedx-vscode-core/src/commands/auth/forceAuthWebLogin.ts

```diff
@@ -22,39 +24,136 @@ import {
 import { getRootWorkspacePath, isSFDXContainerMode } from '../../util';
 import {
   DemoModePromptGatherer,
   SfdxCommandlet,
   SfdxCommandletExecutor,
   SfdxWorkspaceChecker
 } from '../util';
+import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
 import { ForceAuthLogoutAll } from './forceAuthLogout';
 
-import { AuthParams, AuthParamsGatherer } from './authParamsGatherer';
+interface DeviceCodeResponse {
+  user_code: string;
+  device_code: string;
+  interval: number;
+  verification_uri: string;
+}
 
-export class ForceAuthWebLoginExecutor extends SfdxCommandletExecutor<
+export class ForceAuthWebLoginContainerExecutor extends SfdxCommandletExecutor<
   AuthParams
 > {
-  protected showChannelOutput = isSFDXContainerMode();
+  protected showChannelOutput = false;
 
   public build(data: AuthParams): Command {
     const command = new SfdxCommandBuilder().withDescription(
       nls.localize('force_auth_web_login_authorize_org_text')
     );
-    if (isSFDXContainerMode()) {
-      command
-        .withArg('force:auth:device:login')
-        .withLogName('force_auth_device_login');
-    } else {
-      command
-        .withArg('force:auth:web:login')
-        .withLogName('force_auth_web_login');
+
+    command
+      .withArg('force:auth:device:login')
+      .withLogName('force_auth_device_login')
+      .withFlag('--setalias', data.alias)
+      .withFlag('--instanceurl', data.loginUrl)
+      .withArg('--setdefaultusername')
+      .withJson();
+
+    return command.build();
+  }
+
+  public execute(response: ContinueResponse<AuthParams>): void {
+    const startTime = process.hrtime();
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: getRootWorkspacePath(),
+      env: { SFDX_JSON_TO_STDOUT: 'true' }
+    }).execute(cancellationToken);
+    let deviceCodeReceived = false;
+
+    channelService.streamCommandStartStop(execution);
+
+    let stdOut = '';
+    execution.stdoutSubject.subscribe(cliResponse => {
+      stdOut += cliResponse.toString();
+
+      if (!deviceCodeReceived) {
+        const authUrl = this.parseAuthUrlFromStdOut(stdOut);
+
+        if (authUrl) {
+          deviceCodeReceived = true;
+          // open the default browser
+          vscode.env.openExternal(vscode.Uri.parse(authUrl, true));
+        }
+      }
+    });
+
+    execution.processExitSubject.subscribe(() => {
+      this.logMetric(execution.command.logName, startTime);
+    });
+
+    notificationService.reportCommandExecutionStatus(
+      execution,
+      cancellationToken
+    );
+
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private parseAuthUrlFromStdOut(stdOut: string): string | undefined {
```

Comment:
This seems like a good Candidate for a service class/module.  Then it could be unit tested.  Maybe a static method on [AuthUtil.ts](https://github.com/forcedotcom/salesforcedx-vscode/blob/b032cf46357971e91d8e892239f806741e3cdf95/packages/salesforcedx-utils-vscode/src/auth/authUtil.ts#L13). thoughts?

Reply:
Personally I'm not sure it deserves it own class yet. This function is very specific to the device auth and only used in this one execute method at this time. It also sounds like you're gonna rework the auth flows soon? Maybe then would be a better time to evaluate any abstractions that might help.